File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第10章：傅里叶变换/10.3/saliencyMap.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>
using namespace cv;
#include<iostream>
using namespace std;
void fft2Image(InputArray _src, OutputArray _dst);
//求幅度谱 amplitude spectrum
void amplitudeSpectrum(InputArray _srcFFT, OutputArray _dstSpectrum);
//求相位谱 phase spectrum
Mat phaseSpectrum(Mat _srcFFT);
int main(int argc, char*argv[])
{
	//读入图像（灰度化）
	Mat image = imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	if (!image.data)
		return -1;
	imshow("原图", image);
	//转换为 double 类型
	Mat fImage;
	image.convertTo(fImage, CV_64FC1,1.0/255);
	//快速傅里叶变换
	Mat fft2;
	fft2Image(fImage, fft2);
	//幅度谱（又称傅里叶谱）
	Mat amplitude;
	amplitudeSpectrum(fft2, amplitude);
	//对幅度谱进行对数运算
	Mat logAmplitude;
	cv::log(amplitude + 1.0, logAmplitude);
	//均值平滑
	Mat meanLogAmplitude;
	cv::blur(logAmplitude, meanLogAmplitude, Size(3, 3),Point(-1,-1));
	//谱残差
	Mat spectralResidual = logAmplitude - meanLogAmplitude;
	//相位谱
	Mat phase = phaseSpectrum(fft2);
	//余弦谱 cos(phase)
	Mat cosSpectrum(phase.size(), CV_64FC1);
	//正弦谱 sin(phase)
	Mat sinSpectrum(phase.size(), CV_64FC1);
	for (int r = 0; r < phase.rows; r++)
	{
		for (int c = 0; c < phase.cols; c++)
		{
			cosSpectrum.at<double>(r, c) = cos(phase.at<double>(r, c));
			sinSpectrum.at<double>(r, c) = sin(phase.at<double>(r, c));
		}
	}
	//指数运算
	exp(spectralResidual, spectralResidual);
	Mat real = spectralResidual.mul(cosSpectrum);
	Mat imaginary = spectralResidual.mul(sinSpectrum);
	vector<Mat> realAndImag;
	realAndImag.push_back(real);
	realAndImag.push_back(imaginary);
	Mat complex;
	merge(realAndImag, complex);
	//快速傅里叶逆变换
	Mat ifft2;
	dft(complex, ifft2, DFT_COMPLEX_OUTPUT + DFT_INVERSE);
	//傅里叶逆变换的幅度
	Mat ifft2Amp;
	amplitudeSpectrum(ifft2, ifft2Amp);
	//平方运算
	pow(ifft2Amp, 2.0, ifft2Amp);
	//高斯平滑
	GaussianBlur(ifft2Amp, ifft2Amp, Size(11,11), 2.5);
	//显著性显示
	normalize(ifft2Amp, ifft2Amp, 1.0, 0, NORM_MINMAX);
	//提升对比度，进行伽马变换
	pow(ifft2Amp, 0.5, ifft2Amp);
	//数据类型转换
	Mat saliencyMap;
	ifft2Amp.convertTo(saliencyMap, CV_8UC1,255);
	imshow("显著性", saliencyMap);

	waitKey(0);
	return 0;
}
//快速傅里叶变换
void fft2Image(InputArray _src, OutputArray _dst)
{
	//得到Mat类型
	Mat src = _src.getMat();
	//判断位深
	CV_Assert(src.type() == CV_32FC1 || src.type() == CV_64FC1);
	CV_Assert(src.channels() == 1 || src.channels() == 2);
	int rows = src.rows;
	int cols = src.cols;
	//为了进行快速的傅里叶变换，我们经行和列的扩充,找到最合适扩充值
	Mat padded;
	int rPadded = getOptimalDFTSize(rows);
	int cPadded = getOptimalDFTSize(cols);
	//进行边缘扩充,扩充值为零
	copyMakeBorder(src, padded, 0, rPadded - rows, 0, cPadded - cols, BORDER_CONSTANT, Scalar::all(0));
	//快速的傅里叶变换（双通道：用于存储实部 和 虚部）
	dft(padded, _dst, DFT_COMPLEX_OUTPUT);
}
//求幅度谱 amplitude spectrum
void amplitudeSpectrum(InputArray _srcFFT, OutputArray _dstSpectrum)
{
	//判断傅里叶变换是两个通道
	CV_Assert(_srcFFT.channels() == 2);
	//分离通道
	vector<Mat> FFT2Channel;
	split(_srcFFT, FFT2Channel);
	//计算傅里叶变换的幅度谱 sqrt(pow(R,2)+pow(I,2))
	magnitude(FFT2Channel[0],FFT2Channel[1], _dstSpectrum);
}

//求相位谱 phase spectrum
Mat phaseSpectrum(Mat _srcFFT)
{
	//相位谱
	Mat phase;
	phase.create(_srcFFT.size(),CV_64FC1);
	//分离通道
	vector<Mat> FFT2Channel;
	split(_srcFFT, FFT2Channel);
	//计算相位谱
	for (int r = 0; r<phase.rows; r++)
	{
		for (int c = 0; c < phase.cols; c++)
		{
			//实部 
			double real = FFT2Channel[0].at<double>(r, c);
			//虚部
			double imaginary = FFT2Channel[1].at<double>(r, c);
			//计算相位角
			phase.at<double>(r, c) = atan2(imaginary, real);
		}
	}
	return phase;
}

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第10章：傅里叶变换/10.3/saliencyMap.py
# -*- coding: utf-8 -*-
import sys
import numpy as np
import cv2
#快速傅里叶变换
def fft2Image(src):
    #得到行、列
    r,c = src.shape[:2]
    #得到快速傅里叶变换最优
    rPadded = cv2.getOptimalDFTSize(r)
    cPadded = cv2.getOptimalDFTSize(c)
    #边缘扩充，下边缘和右边缘扩充值为零
    fft2 = np.zeros((rPadded,cPadded,2),np.float32)
    fft2[:r,:c,0]=src
    #快速傅里叶变换
    cv2.dft(fft2,fft2,cv2.DFT_COMPLEX_OUTPUT)
    return fft2
 #傅里叶幅度谱
def amplitudeSpectrum(fft2):
    #求幅度
    real,Imag = cv2.split(fft2)
    amplitude = cv2.magnitude(real,Imag)
    #amplitude = cv2.magnitude(fft2[:,:,0],fft2[:,:,1])
    return amplitude
#幅度谱的灰度级显示
def graySpectrum(amplitude):
    #对比度拉伸
    #cv2.log(amplitude+1.0,amplitude)
    amplitude = np.log(amplitude+1.0)
    #归一化,傅里叶谱的灰度级显示
    spectrum = np.zeros(amplitude.shape,np.float32)
    cv2.normalize(amplitude,spectrum,0,1,cv2.NORM_MINMAX)
    return spectrum
#相位谱
def phaseSpectrum(fft2):
    #得到行数、列数
    rows,cols = fft2.shape[:2]
    #计算相位角
    phase = np.arctan2(fft2[:,:,1],fft2[:,:,0])
    #显示该相位谱时，首先需要将相位角转换为 [ -180 , 180]
    #spectrum = phase/math.pi*180
    return phase    
#主函数
if __name__ =="__main__":
    if len(sys.argv) > 1:
        image = cv2.imread(sys.argv[1],cv2.IMREAD_GRAYSCALE)
    else:
        print "Usge:python saliencyMap.py imageFile"
    #第一步：计算图像的快速傅里叶变换
    fft2 = fft2Image(image)
    #第二步：计算傅里叶幅度谱的灰度级
    #求幅度谱
    amplitude = amplitudeSpectrum(fft2)
    #幅度谱的灰度级
    logAmplitude = graySpectrum(amplitude)
    #第三步：计算相位
    phase = phaseSpectrum(fft2)
    #余弦谱（用于计算实部）
    cosSpectrum = np.cos(phase)
    #正弦谱（用于计算虚部）
    sinSectrum = np.sin(phase)
    #第四步：计算残差（Spectral Residual）
    #对幅度谱的灰度级进行均值平滑
    meanLogAmplitude = cv2.boxFilter(logAmplitude,cv2.CV_32FC1,(3,3))
    #残差
    spectralResidual = logAmplitude - meanLogAmplitude
    #第五步：计算傅里叶逆变换,显著性
    #残差的指数
    expSR = np.exp(spectralResidual)
    #分别计算实部和虚部
    real = expSR*cosSpectrum
    imaginary = expSR*sinSectrum
    #合并实部和虚部
    com = np.zeros((real.shape[0],real.shape[1],2),np.float32)
    com[:,:,0]=real
    com[:,:,1]=imaginary
    #逆变换
    ifft2=np.zeros(com.shape,np.float32)
    cv2.dft(com,ifft2,cv2.DFT_COMPLEX_OUTPUT+cv2.DFT_INVERSE)
    #显著性
    saliencymap = np.power(ifft2[:,:,0],2)+np.power(ifft2[:,:,1],2)
    #对显著性进行高斯平滑
    saliencymap = cv2.GaussianBlur(saliencymap,(11,11),2.5)
    #显示检测到的显著性
    #cv2.normalize(saliencymap,saliencymap,0,1,cv2.NORM_MINMAX)
    saliencymap  = saliencymap/np.max(saliencymap)
    #提高对比度，进行伽马变换
    saliencymap = np.power(saliencymap,0.5)
    saliencymap = np.round(saliencymap*255)
    saliencymap = saliencymap.astype(np.uint8)
    cv2.imshow("saliencymap",saliencymap)
    cv2.imwrite("saliencymap.jpg",saliencymap)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
        